pkg/velerorunner/registry.go

```diff
@@ -0,0 +1,526 @@
+		err = client.Create(context.TODO(), newDeploymentConfig)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	registryResources.DeploymentConfig = foundDeploymentConfig
+	if !t.equalsRegistryDeploymentConfig(newDeploymentConfig, foundDeploymentConfig) {
+		t.updateRegistryDeploymentConfig(foundDeploymentConfig, name, srcName)
+		err = client.Update(context.TODO(), foundDeploymentConfig)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Build a Registry DeploymentConfig as desired for the source cluster.
+func (t *Task) buildRegistryDeploymentConfig(name, srcName string) (*appsv1API.DeploymentConfig, error) {
+	labels := t.PlanResources.MigPlan.GetCorrelationLabels()
+	labels["app"] = name
+	deploymentconfig := &appsv1API.DeploymentConfig{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:    labels,
+			Name:      name,
+			Namespace: VeleroNamespace,
+		},
+	}
+	err := t.updateRegistryDeploymentConfig(deploymentconfig, name, srcName)
+	return deploymentconfig, err
+}
+
+// Update a Registry DeploymentConfig as desired for the specified cluster.
+func (t *Task) updateRegistryDeploymentConfig(deploymentconfig *appsv1API.DeploymentConfig, name, srcName string) error {
+	deploymentconfig.Spec = appsv1API.DeploymentConfigSpec{
+		Replicas: 1,
+		Selector: map[string]string{
+			"app":              name,
+			"deploymentconfig": name,
+		},
+		Strategy: appsv1API.DeploymentStrategy{Resources: kapi.ResourceRequirements{}},
+		Template: &kapi.PodTemplateSpec{
+			ObjectMeta: metav1.ObjectMeta{
+				CreationTimestamp: metav1.Time{},
+				Labels: map[string]string{
+					"app":              name,
+					"deploymentconfig": name,
+				},
+			},
+			Spec: kapi.PodSpec{
+				Containers: []kapi.Container{
+					kapi.Container{
+						Env: []kapi.EnvVar{
+							kapi.EnvVar{
+								Name:  "REGISTRY_STORAGE",
+								Value: "s3",
+							},
+							kapi.EnvVar{
+								Name: "REGISTRY_STORAGE_S3_ACCESSKEY",
+								ValueFrom: &kapi.EnvVarSource{
+									SecretKeyRef: &kapi.SecretKeySelector{
+										LocalObjectReference: kapi.LocalObjectReference{Name: name},
+										Key:                  "access_key",
+									},
+								},
+							},
+							kapi.EnvVar{
+								Name:  "REGISTRY_STORAGE_S3_BUCKET",
+								Value: t.PlanResources.MigStorage.Spec.BackupStorageConfig.AwsBucketName,
+							},
+							kapi.EnvVar{
+								Name:  "REGISTRY_STORAGE_S3_REGION",
+								Value: t.PlanResources.MigStorage.Spec.BackupStorageConfig.AwsRegion,
+							},
+							kapi.EnvVar{
+								Name:  "REGISTRY_STORAGE_S3_ROOTDIRECTORY",
+								Value: "/" + srcName,
+							},
+							kapi.EnvVar{
+								Name: "REGISTRY_STORAGE_S3_SECRETKEY",
+								ValueFrom: &kapi.EnvVarSource{
+									SecretKeyRef: &kapi.SecretKeySelector{
+										LocalObjectReference: kapi.LocalObjectReference{Name: name},
+										Key:                  "secret_key",
+									},
+								},
+							},
+						},
+						Image: "registry:2",
```

Comment:
Is this going to start a container running an image called `registry` with tag `2`? Can we expect `image:2` to exist on our clusters?

Reply:
We can because of the ImageStream we just created above which matches this.